### PR TITLE
Fix GetBundleContext when using DS

### PR DESCRIPTION
**Pull request by bwestcot**
For the resolution of https://github.com/CppMicroServices/CppMicroServices/issues/491

### CODE REVIEW

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

File: text/0000-Declarative-Services-GetBundleContext.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+- Start Date: 2023-09-18
+- RFC PR: (in a subsequent commit to the PR, fill me with the PR's URL)
+- CppMicroServices Issue: https://github.com/CppMicroServices/CppMicroServices/issues/491
+
+# Supporting GetBundleContext inside Declarative Services
+
+## Summary
+
+Currently, the freestanding `GetBundleContext` function generally does not work when used inside a declarative service. This is for two reasons: the necessary data is usually not present in the bundle object, and even when present, that data is often not initialized. This RFC is for the resolution of these two deficiencies.
```
[the remark this answers is not in the document]
Comment (bwestcot) [reply text/0000-Declarative-Services-GetBundleContext.md:9]: Based on my understanding and points 1 and 2 under "Detailed Design", it *can* work, if:
1. The bundle developer uses the CPPMICROSERVICES_INITIALIZE_BUNDLE macro, creating the data storage
2. The bundle is set to load immediately
3. The bundle is specified as having a bundle activator, causing the immediate loader to initialize the data storage

Is this incorrect?

File: text/0000-Declarative-Services-GetBundleContext.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+1. Use of the macro that provides data storage for `GetBundleContext` is discouraged when using Declarative Services, but its functionality has not been replaced.
+2. A logic error in the immediate bundle loader means that this data storage is not initialized unless `"bundle.activator"` is set to `true` in the bundle manifest.
+3. The lazy bundle loader, which is the default loader, does not initialize this data storage under any circumstances.
+
+The data storage and its accessor functions that allow `GetBundleContext` to work are provided when the user adds the `CPPMICROSERVICES_INITIALIZE_BUNDLE` macro to their bundle code (`BundleInitialization.h`). While this is required without DS, our current documentation shows it not being used with DS. Because the DS Code Generation Tool already provides a facility for inserting code into the bundle, the tool will be modified to insert the necessary code, effectively providing the contents of the `CPPMICROSERVICES_INITIALIZE_BUNDLE` macro for the developer automatically.
```
Comment (carneyweb) [inline text/0000-Declarative-Services-GetBundleContext.md:23]: Same issue here. Here's the definition of the macro:

    #    define CPPMICROSERVICES_INITIALIZE_BUNDLE                                                                       \
        std::atomic<cppmicroservices::BundleContextPrivate*> US_CTX_INS(US_BUNDLE_NAME) {};                          \
                                                                                                                     \
        extern "C" cppmicroservices::BundleContextPrivate* US_GET_CTX_FUNC(US_BUNDLE_NAME)();                        \
        extern "C" cppmicroservices::BundleContextPrivate* US_GET_CTX_FUNC(US_BUNDLE_NAME)()                         \
        {                                                                                                            \
            return US_CTX_INS(US_BUNDLE_NAME).load();                                                                \
        }                                                                                                            \
                                                                                                                     \
        extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)(cppmicroservices::BundleContextPrivate * ctx); \
        extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)(cppmicroservices::BundleContextPrivate * ctx)  \
        {                                                                                                            \
            US_CTX_INS(US_BUNDLE_NAME).store(ctx);                                                                   \
        }

Where is data storage allocated? I see he declaration of an atomic BundleContextPrivate*, and a function to get its value and set its value. No allocations, no storage provided.

**Review by carneyweb — COMMENTED** [review]
I think this looks good, except for the references to storage allocation. Those need to be reworded, unless I'm completely misunderstanding what I'm reading (which is entirely possible :-)).

File: text/0000-Declarative-Services-GetBundleContext.md (hunk 2 above)
Comment (bwestcot) [reply text/0000-Declarative-Services-GetBundleContext.md:23]: I got tripped up by the same thing when I first read this code. The first line declares a brace-initialized global variable - the US_CTX_INS(US_BUNDLE_NAME) macros expand to a variable name. Though the parens and braces make it look like an empty function at first glance. Note the extra pair of parens for the function declarations.

Comment (carneyweb) [reply text/0000-Declarative-Services-GetBundleContext.md:23]: Yep. I got that. The first line declares a global variable that is just a pointer to an object. Nothing is allocated. The pointer (which points to the allocated memory) must be set with a call to the US_SET_CTX_FUNC().

Comment (bwestcot) [reply text/0000-Declarative-Services-GetBundleContext.md:23]: I wrote this thinking about the std::atomic object, whose purpose is storing some data we want later (8 bytes of pointer). Though conceptually we care more about the BundleContextPrivate it points to. I'll adjust the wording to focus on that instead.

Comment (carneyweb) [reply text/0000-Declarative-Services-GetBundleContext.md:23]: Perfect. :-)

File: text/0000-Declarative-Services-GetBundleContext.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bwestcot) [reply text/0000-Declarative-Services-GetBundleContext.md]: Confirmed by testing

[the remark this answers is not in the document]
Comment (jeffdiclemente) [reply text/0000-Declarative-Services-GetBundleContext.md]: that's far. I would still like to see a re-wording that more explicitly details the conditions needed for GetBundleContext to behave correctly. The enumerated list you provided in your comments above is the level of detail that I think is needed in the summary.

[the remark this answers is not in the document]
Comment (bwestcot) [reply text/0000-Declarative-Services-GetBundleContext.md]: I've revised the summary to include this list.

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

File: text/0000-Declarative-Services-GetBundleContext.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+The immediate bundle loader already has the necessary logic for initializing the context pointer, but it is guarded by a conditional that it should not be. The resolution for this component is to simply pull the initialization outside of that conditional. (`BundlePrivate::Start0`)
+
+As the lazy bundle loader completely lacks initialization code, it will need to be copied from the immediate bundle loader and be placed such that it runs once, after the bundle is loaded into memory but before the bundle activator is run. (`scrimpl::GetComponentCreatorDeletors`)
+
+If initialization fails, for example because the symbols do not exist in the bundle, an error will be logged and initialization will continue. The bundle signature check should prevent corrupted bundles from loading. Not failing will allow existing DS bundles that were built before this fix to still be loaded and used.
+
+Tests for the DS Code Generation changes will be added to `SCRCodegenTests`, and tests for the bundle loaders will be added to `usDeclarativeServicesTests`.
```
[the remark this answers is not in the document]
Comment (bwestcot) [reply text/0000-Declarative-Services-GetBundleContext.md]: from call - no, changes are not significant

**Review by jeffdiclemente — APPROVED** [review]

**Pull request merged**